x/concentrated-liquidity/math/tick.go

```diff
@@ -168,51 +168,45 @@ func powTenBigDec(exponent int64) osmomath.BigDec {
+	var geoSpacing *tickExpIndexData
 	if price.GT(sdkOneDec) {
-		for currentPrice.LT(price) {
-			currentAdditiveIncrementInTicks = powTenBigDec(exponentAtCurrentTick)
-			maxPriceForCurrentAdditiveIncrementInTicks := osmomath.NewBigDec(geometricExponentIncrementDistanceInTicks).Mul(currentAdditiveIncrementInTicks)
-			currentPrice = currentPrice.Add(maxPriceForCurrentAdditiveIncrementInTicks.SDKDec())
-			exponentAtCurrentTick = exponentAtCurrentTick + 1
-			ticksPassed = ticksPassed + geometricExponentIncrementDistanceInTicks
+		index := 0
+		geoSpacing = tickExpCache[int64(index)]
+		for geoSpacing.maxPrice.LT(price) {
+			index += 1
+			geoSpacing = tickExpCache[int64(index)]
 		}
 	} else {
-		// We must decrement the exponentAtCurrentTick by one when traversing negative ticks in order to constantly step up in precision when going further down in ticks
-		// Otherwise, from tick 0 to tick -(geometricExponentIncrementDistanceInTicks), we would use the same exponent as the exponentAtPriceOne
-		exponentAtCurrentTick := exponentAtPriceOne - 1
-		for currentPrice.GT(price) {
-			currentAdditiveIncrementInTicks = powTenBigDec(exponentAtCurrentTick)
-			maxPriceForCurrentAdditiveIncrementInTicks := osmomath.NewBigDec(geometricExponentIncrementDistanceInTicks).Mul(currentAdditiveIncrementInTicks)
-			currentPrice = currentPrice.Sub(maxPriceForCurrentAdditiveIncrementInTicks.SDKDec())
-			exponentAtCurrentTick = exponentAtCurrentTick - 1
-			ticksPassed = ticksPassed - geometricExponentIncrementDistanceInTicks
+		index := -1
+		geoSpacing = tickExpCache[int64(index)]
+		for geoSpacing.initialPrice.GT(price) {
+			index -= 1
+			geoSpacing = tickExpCache[int64(index)]
 		}
 	}
 
-	// Determine how many ticks we have passed in the exponentAtCurrentTick (in other words, the incomplete geometricExponent above)
-	ticksToBeFulfilledByExponentAtCurrentTick := osmomath.BigDecFromSDKDec(price.Sub(currentPrice)).Quo(currentAdditiveIncrementInTicks)
-
-	// Finally, add the ticks we have passed from the completed geometricExponent values, as well as the ticks we have passed in the current geometricExponent value
-	tickIndex = ticksPassed + ticksToBeFulfilledByExponentAtCurrentTick.SDKDec().RoundInt64()
+	// We know were between (geoSpacing.initialPrice, geoSpacing.endPrice)
+	// The number of ticks that need to be filled by our current spacing is
+	// (price - geoSpacing.initialPrice) / geoSpacing.additiveIncrementPerTick
+	priceInThisExponent := osmomath.BigDecFromSDKDec(price.Sub(geoSpacing.initialPrice))
+	ticksFilledByCurrentSpacing := priceInThisExponent.Quo(geoSpacing.additiveIncrementPerTick)
+	// we get the bucket index by:
+	// * taking the bucket index of the smallest price in this tick
+	// * adding to it the number of ticks "completely" filled by the current spacing
+	// the latter is the truncation of the division above
+	// TODO: This should be rounding down?
+	tickIndex = geoSpacing.initialTick + ticksFilledByCurrentSpacing.SDKDec().RoundInt64()
```

Comment:
If we do roundInt64, we are defining a bucket `b` to be the range `TickToPrice(b - .5)`, `TickToPrice(b + .5)`

If we want the definition (which I think we intend), we should be doing bucket `b` is the range `TickToPrice(b)` to `TickToPrice(b + 1)`. That would be rounding down. If we agree, lets change that in a subsequent PR

Reply:
The definition makes sense to me. However, I don't understand why many inverse relationship tests break if we make this change.